Library/ViewModels/CreatePasswordViewModel.swift

```diff
@@ -82,6 +87,32 @@ CreatePasswordViewModelInputs, CreatePasswordViewModelOutputs {
 
     self.saveButtonIsEnabled = formIsValid
 
+    let autoSaveSignal = self.saveButtonIsEnabled
+      .takeWhen(self.newPasswordConfirmationDidReturnProperty.signal)
+      .filter { isTrue($0) }
+      .ignoreValues()
+
+    let triggerSaveAction = Signal.merge(autoSaveSignal, self.saveButtonTappedProperty.signal)
+
+    let createPasswordEvent = combinedPasswords
+      .takeWhen(triggerSaveAction)
+      .map { CreatePasswordInput(password: $0.0, passwordConfirmation: $0.1) }
+      .switchMap { input in
+        AppEnvironment.current.apiService.createPassword(input: input)
+          .ksr_delay(AppEnvironment.current.apiDelayInterval, on: AppEnvironment.current.scheduler)
+          .materialize()
+    }
+
+    self.createPasswordSuccess = createPasswordEvent.values().ignoreValues()
+
+    self.activityIndicatorShouldShow = Signal.merge(
+      triggerSaveAction.signal.mapConst(true),
+      self.createPasswordSuccess.mapConst(false)
+    )
+
+    self.dismissKeyboard = Signal.merge(self.saveButtonTappedProperty.signal,
```

Comment:
Looks like there are couple events that are repeated here...wonder if we can simplify this a bit.

```
// This is being called when user taps 'Save' or `Done'
let submitFormEvent = Signal.merge(
  self.newPasswordConfirmationDidReturnProperty.signal,
  self.saveButtonTappedProperty.signal
)

// So the following action isn't only concerned about 'Done' key but both 'Done' & 'Save'
// 'autoSaveSignal' becomes 'saveAction'
// allows us to remove 'triggerSaveAction' altogether
let saveAction = formIsValid
  .takeWhen(submitFormEvent)
  .filter(isTrue)
  .ignoreValues()

let createPasswordEvent = combinedPasswords
  .takeWhen(saveAction)
  .map { CreatePasswordInput(password: $0.0, passwordConfirmation: $0.1) }
  .ignoreValues()
  ... // call graph here //

self.createPasswordSuccess = createPasswordEvent

self.activityIndicatorShouldShow = Signal.merge(
  saveAction.signal.mapConst(true),
  self.createPasswordSuccess.mapConst(false)
)

self.dismissKeyboard = submitFormEvent
```

Also I haven't really seen us naming variables with as `somethingSignal` but rather `somethingEvent` or `somethingAction` so wonder if for consistency we should not use `autoSaveSignal`?

Reply:
I agree, renamed it.

Reply:
Also not going to `ignoreValues()` of the `createPasswordEvent` because I'll need the error description later.